File: backend/log_redactor.py
"""§G-2 — Python 로깅 redaction filter.

logging.Filter 서브클래스. record.msg 와 record.args 의 문자열들을 정규식으로 마스킹.
APP_TOKEN, GEMINI_API_KEY, naver_pw, X-App-Token 등이 stdout/파일 로그에 누설되지 않게 한다.
"""
from __future__ import annotations

import logging
import re

_PATTERNS: list[tuple[re.Pattern[str], str]] = [
    (re.compile(r"((?:APP_TOKEN|APP_SESSION_TOKEN|GEMINI_API_KEY)=)([^\s'\"]+)"), r"\1***"),
    (re.compile(r"(X-App-(?:Token|Session)\s*[:=]\s*)([^\s'\",}]+)", re.IGNORECASE), r"\1***"),
    (re.compile(r"(Authorization\s*[:=]\s*)([^\s'\",}]+)", re.IGNORECASE), r"\1***"),
    (re.compile(r'("?naver_pw"?\s*[:=]\s*)("?)([^,"\n}]+)'), r"\1\2***"),
    (re.compile(r'("?naver_pw_encrypted"?\s*[:=]\s*)("?)([A-Za-z0-9+/=]+)'), r"\1\2***"),
]


def _redact(text: str) -> str:
    out = text
    for pat, repl in _PATTERNS:
        out = pat.sub(repl, out)
    return out


class RedactingFilter(logging.Filter):
    def filter(self, record: logging.LogRecord) -> bool:
        # msg 가 문자열이면 마스킹. 다른 타입이면 건드리지 않음.
        if isinstance(record.msg, str):
            record.msg = _redact(record.msg)
        # args 안의 문자열도 마스킹.
        if record.args:
            new_args = []
            if isinstance(record.args, tuple):
                for arg in record.args:
                    new_args.append(_redact(arg) if isinstance(arg, str) else arg)
                record.args = tuple(new_args)
            elif isinstance(record.args, dict):
                record.args = {
                    k: (_redact(v) if isinstance(v, str) else v) for k, v in record.args.items()
                }
        return True

File: electron/src/log-redactor.ts
// §G-2 — 로그에서 토큰/키/비밀번호 마스킹.
//
// electron-log 의 transform hook 에 끼워, file/console transport 모두에 적용.
// 새로운 토큰/키 종류를 도입할 때마다 패턴을 보강.

const PATTERNS: Array<[RegExp, string]> = [
  // KEY=value
  [/((?:APP_TOKEN|APP_SESSION_TOKEN|GEMINI_API_KEY)=)([^\s'"]+)/g, "$1***"],
  // 헤더 X-App-Token: xxx
  [/(X-App-(?:Token|Session)\s*[:=]\s*)([^\s'",}]+)/gi, "$1***"],
  // Authorization: Bearer xxx
  [/(Authorization\s*[:=]\s*)([^\s'",}]+)/gi, "$1***"],
  // naver_pw 필드
  [/("?naver_pw"?\s*[:=]\s*)("?)([^,"\n}]+)/g, '$1$2***'],
  // naver_pw_encrypted 필드 — ciphertext 자체도 마스킹 (길이도 노출 안 함)
  [/("?naver_pw_encrypted"?\s*[:=]\s*)("?)([A-Za-z0-9+/=]+)/g, '$1$2***'],
  // base64 plaintext 가 우연히 보일 가능성 차단 (가장 마지막에 적용되지 않으니 신중)
];

export function redact(input: unknown): unknown {
  if (typeof input !== "string") return input;
  let out = input;
  for (const [re, repl] of PATTERNS) {
    out = out.replace(re, repl);
  }
  return out;
}

/** electron-log hook 으로 쓸 함수. message.data 의 각 인자를 redact 후 반환. */
export function redactTransform<T extends { data: unknown[] }>(message: T): T {
  message.data = message.data.map(redact);
  return message;
}
